src/common/types/value/value.cpp

```diff
@@ -732,19 +731,25 @@ void Value::copyFromUnion(const uint8_t* kuUnion) {
     // For union dataType, only one member can be active at a time. So we don't need to copy all
     // union fields into value.
     auto activeFieldIdx = UnionType::getInternalFieldIdx(*(union_field_idx_t*)unionValues);
-    auto childValue = children[0].get();
-    childValue->dataType = childrenTypes[activeFieldIdx]->copy();
+    // Create default value now that we know the active field
+    auto childValue = Value::createDefaultValue(*childrenTypes[activeFieldIdx]);
     auto curMemberIdx = 0u;
     // Seek to the current active member value.
     while (curMemberIdx < activeFieldIdx) {
         unionValues += storage::StorageUtils::getDataTypeSize(*childrenTypes[curMemberIdx]);
         curMemberIdx++;
     }
     if (NullBuffer::isNull(unionNullValues, activeFieldIdx)) {
-        childValue->setNull(true);
+        childValue.setNull(true);
+    } else {
+        childValue.setNull(false);
+        childValue.copyFromRowLayout(unionValues);
+    }
+    if (children.empty()) {
+        children.push_back(std::make_unique<Value>(std::move(childValue)));
+        childrenSize = 1;
     } else {
-        childValue->setNull(false);
-        childValue->copyFromRowLayout(unionValues);
+        children[0] = std::make_unique<Value>(std::move(childValue));
```

Comment:
I am a bit confused there?
When will the children by empty? Since you always push a value to the `createDefaultValue` function:
```
ase LogicalTypeID::UNION: {
        // We can't create a default value for the union since the
        // selected variant is runtime information. Default value
        // is initialized when copying (see Value::copyFromUnion).
        return Value(dataType.copy(), std::vector<std::unique_ptr<Value>>{});
```